### ISSUE TITLE
Xerces: New version 3.1.0 in CMSSW_8_1_X

### ISSUE THREAD

**Issue by iahmad-khan** [cms-sw/cmssw#14084]
Requested: by @ianna 
Requires:  mcdb 1.0.3
Requires:  tkonlinesw 4.1.0-1

The utm also need an update , the errors from utm log:

```
In file included from ../xsd/cxx/parser/xerces/elements.txx:16:0,
                 from ../xsd/cxx/parser/xerces/elements.hxx:449,
                 from ../tmXsd/gen-xsd-type/simple-types-pskel.hh:80,
                 from ../tmXsd/gen-xsd-type/simple-types-parser.hh:7,
                 from reader.cc:9:
../xsd/cxx/xml/sax/std-input-source.hxx: In member function 'virtual xercesc_3_1::BinInputStream* xsd::cxx::xml::sax::std_input_source::makeStream
() const':
../xsd/cxx/xml/sax/std-input-source.hxx:156:44: error: invalid new-expression of abstract class type 'xsd::cxx::xml::sax::std_input_stream'
             return new std_input_stream (is);
                                            ^
../xsd/cxx/xml/sax/std-input-source.hxx:24:15: note:   because the following virtual functions are pure within 'xsd::cxx::xml::sax::std_input_stre
am':
         class std_input_stream: public xercesc::BinInputStream
               ^
In file included from ../xsd/cxx/xml/sax/std-input-source.hxx:14:0,
                 from ../xsd/cxx/parser/xerces/elements.txx:16,
                 from ../xsd/cxx/parser/xerces/elements.hxx:449,
                 from ../tmXsd/gen-xsd-type/simple-types-pskel.hh:80,
                 from ../tmXsd/gen-xsd-type/simple-types-parser.hh:7,
                 from reader.cc:9:
/build/build/iahmad/build/slc6_amd64_gcc530/external/xerces-c/3.1.0/include/xercesc/util/BinInputStream.hpp:67:26: note:        virtual const XMLC
h* xercesc_3_1::BinInputStream::getContentType() const
     virtual const XMLCh* getContentType() const = 0;
                          ^
In file included from ../xsd/cxx/parser/xerces/elements.txx:16:0,
                 from ../xsd/cxx/parser/xerces/elements.hxx:449,
                 from ../tmXsd/gen-xsd-type/simple-types-pskel.hh:80,
                 from ../tmXsd/gen-xsd-type/simple-types-parser.hh:7,
                 from ../tmXsd/reader.hh:19,
                 from tree2table.cc:13:
../xsd/cxx/xml/sax/std-input-source.hxx: In member function 'virtual xercesc_3_1::BinInputStream* xsd::cxx::xml::sax::std_input_source::makeStream() const':
../xsd/cxx/xml/sax/std-input-source.hxx:156:44: error: invalid new-expression of abstract class type 'xsd::cxx::xml::sax::std_input_stream'
             return new std_input_stream (is);
                                            ^
../xsd/cxx/xml/sax/std-input-source.hxx:24:15: note:   because the following virtual functions are pure within 'xsd::cxx::xml::sax::std_input_stream':
         class std_input_stream: public xercesc::BinInputStream
               ^
In file included from ../xsd/cxx/xml/sax/std-input-source.hxx:14:0,
                 from ../xsd/cxx/parser/xerces/elements.txx:16,
                 from ../xsd/cxx/parser/xerces/elements.hxx:449,
                 from ../tmXsd/gen-xsd-type/simple-types-pskel.hh:80,
                 from ../tmXsd/gen-xsd-type/simple-types-parser.hh:7,
                 from ../tmXsd/reader.hh:19,
                 from tree2table.cc:13:
/build/build/iahmad/build/slc6_amd64_gcc530/external/xerces-c/3.1.0/include/xercesc/util/BinInputStream.hpp:67:26: note:        virtual const XMLCh* xercesc_3_1::BinInputStream::getContentType() const
     virtual const XMLCh* getContentType() const = 0;
                          ^
Makefile:86: recipe for target 'reader.o' failed
make[1]: *** [reader.o] Error 1
make[1]: *** Waiting for unfinished jobs....
Makefile:86: recipe for target 'tree2table.o' failed
make[1]: *** [tree2table.o] Error 1

```

**Comment by smuzaffar** [comment]
assign externals
@mulhearn and @tmatsush

**Comment by tmatsush** [comment]
A new tag, r44601-xsd330, has been created to be compatible with 81X with Xerces-3.1
Makefile has been updated to exit with error code as well.

Also, another tag has been created for 80X with Xerces-2.8 and will create another issue for this update.

**Comment by iahmad-khan** [comment]
I have tried it with 81X using Xerces-3.1 , but it still results in errors:
```
/xsd/cxx/zc-istream.txx:82:45:   required from 'xsd::cxx::zc_istream<C>::zc_istream(const xsd::cxx::ro_string<C>&) [with C = char]'
../../xsd/cxx/tree/parsing/unsigned-int.hxx:69:30:   required from 'static unsigned int xsd::cxx::tree::traits<unsigned int, C, (xsd::cxx::tree::schema_type::value)0u>::create(const std::basic_string<_CharT>&, const xercesc_3_1::DOMElement*, xsd::cxx::tree::flags, xsd::cxx::tree::container*) [with C = char; xsd::cxx::tree::container = xsd::cxx::tree::_type]'
../../xsd/cxx/tree/parsing/unsigned-int.hxx:44:23:   required from 'static unsigned int xsd::cxx::tree::traits<unsigned int, C, (xsd::cxx::tree::schema_type::value)0u>::create(const xercesc_3_1::DOMElement&, xsd::cxx::tree::flags, xsd::cxx::tree::container*) [with C = char; xsd::cxx::tree::container = xsd::cxx::tree::_type]'
complex-bin.cc:242:63:   required from here
../../xsd/cxx/zc-istream.txx:35:12: error: 'setg' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]

```

**Comment by tmatsush** [comment]
Sorry, I was testing against xerces-c-3.0
Is there any testing environment for xerces-c-3.1 ?

**Comment by davidlt** [comment]
Download and build locally for testing 3.1.3. I see Fedora 23 is also shipping 3.1.3, thus it's as easy as installing xerces-c-devel package on Fedora 23 or 24.

**Comment by smuzaffar** [comment]
@iahmad-khan, can upload the cmsdist with Xerces-3.1 changes to your github repository so that @tmatsush can clone and use it for building utm?

**Comment by iahmad-khan** [comment]
Changes uploaded to:
https://github.com/iahmad-khan/cmsdist/tree/IB/CMSSW_8_1_X/gcc530

**Comment by smuzaffar** [comment]
@tmatsush, you can checkout @iahmad-khan branch and try building UTM using pkgtools/cmsBuild. Let us know if you need help using pkgtools

**Comment by tmatsush** [comment]
@smuzaffar, thanks. Please give me some instruction for building UTM with pkgtools.

**Comment by smuzaffar** [comment]
Detail instruction on using pkgtools for building externals are available here http://cms-sw.github.io/cmsdist/ (need some cleanup though). In simple you can just do

 * get cmsdist and pkgtools
```
git clone -b IB/CMSSW_8_1_X/gcc530 git@github.com:iahmad-khan/cmsdist.git
git clone -b V00-22-XX git@github.com:cms-sw/pkgtools
```
 * edit cmsdist/utm.spec and update the utm tag in first list to match the one which should be build using xerces-c 3.1.0

* build utm RPM
```
./pkgtools/cmsBuild -c cmsdist -a slc6_amd64_gcc530 -j 8 build utm
```

if every thing goes fine then it will install utm in slc6_amd64_gcc530/external/utm/version

**Comment by tmatsush** [comment]
Thank you for the instruction.
A new tag, r45210-xsd330-patch, has been prepared which can be compiled with the above instruction.

**Comment by smuzaffar** [comment]
+externals
